tests/grid/grid_tools_guess_pt_owner_02.cc

```diff
@@ -0,0 +1,201 @@
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// test for GridTools::guess_point_owner, tree version:
+// simulating different number of processes with different
+// number of bounding boxes
+
+#include <deal.II/base/bounding_box.h>
+#include <deal.II/base/point.h>
+
+#include <deal.II/grid/grid_tools.h>
+
+#include <tuple>
+
+#include "../tests.h"
+
+template <int spacedim>
+void
+test_point_owner(unsigned int n_procs)
+{
+  // Step 1: creating the vector of bounding boxes
+  // We want to simulate n_procs bounding boxes:
+  // each process of rank rk has rk + 1 bounding boxes
+  // the bounding boxes describe a mesh which is a
+  // segment (in dim 1), a rectangle (in dim 2),
+  // a parallelogram (in dim 3)
+  // and using the center point of each bounding box
+
+  std::vector<std::pair<BoundingBox<spacedim>, unsigned int>> global_bboxes;
+
+  unsigned int tot_bbox = 0;
+
+  for (unsigned int rk = 0; rk < n_procs; ++rk)
+    for (unsigned int box = 0; box < rk; ++box)
+      {
+        std::pair<Point<spacedim>, Point<spacedim>> boundaries;
+        boundaries.first[0]  = tot_bbox;
+        boundaries.second[0] = tot_bbox + 1;
+        for (int i = 1; i < spacedim; i++)
+          boundaries.second[i] = 1;
+
+        BoundingBox<spacedim> new_box(boundaries);
+        global_bboxes.emplace_back(std::make_pair(new_box, rk));
+        ++tot_bbox;
+      }
+
+  // Creating the vector of center points
+  std::vector<Point<spacedim>> points(tot_bbox);
+  for (unsigned int i = 0; i < tot_bbox; ++i)
+    {
+      points[i][0] = i + 0.5;
+      for (unsigned int j = 1; j < spacedim; j++)
+        points[i][j] = 0.5;
+    }
+
+  // Step 2: building the rtree
+  RTree<std::pair<BoundingBox<spacedim>, unsigned int>> covering_rtree(
+    global_bboxes.begin(), global_bboxes.end());
+
+  // Step 2: testing the function
+  auto output_tp = GridTools::guess_point_owner(covering_rtree, points);
+
+  bool test_passed = true;
+
+  if (std::get<2>(output_tp).size() != 0)
+    {
+      test_passed = false;
+      deallog << " Test 1 failed: multiple owners" << std::endl;
+    }
+  else
+    {
+      // Check the points are all in the correct rank
+      unsigned int tot_pt = 0;
+      for (unsigned int rk = 0; rk < n_procs; ++rk)
+        {
+          const auto & rk_points = std::get<0>(output_tp)[rk];
+          unsigned int first_el  = rk * (rk + 1) / 2 - 1;
+          unsigned int last_el   = (rk + 1) * (rk + 2) / 2;
+          for (unsigned int box = 0; box < rk; ++box)
+            {
+              if (std::find(rk_points.begin() + first_el,
+                            rk_points.begin() + last_el,
+                            tot_pt) == rk_points.end())
```

Comment:
@GivAlz you need to help me with this test. In the reference output file, there is one failure reported about this search:
```
DEAL::Simulating 5 processes
DEAL::Point 2 not found in rank 2
DEAL::TEST FAILED
```
We're having trouble with this one on some of our testers, e.g. this one:
https://cdash.kyomu.43-1.org/testDetails.php?test=64971275&build=9779
Even after looking at this test for almost half an hour, I must admit that I don't understand why you limit `std::find` to `rk_points.begin()+first_el, rk_points.begin()+last_el`. We are are here looking at the output of the rank `rk` through `std::get<0>(output_tp)[rk];` in line 89, so I would have assumed that this already is the restricted search. Why is there an additional limit `first_el` and `last_el`, which I would assume only should be there if we make the search in the global array, not in the localized one. What do I miss?

Reply:
Now that I'm looking more closely at it, this search probably accesses invalid memory because `first_el` and `last_el` will point to invalid memory. Shouldn't this read:
```
          const auto & rk_points = std::get<0>(output_tp)[rk];
          for (unsigned int box = 0; box < rk; ++box)
            {
              if (std::find(rk_points.begin(),
                            rk_points.end(),
                            tot_pt) == rk_points.end())
```
Please confirm, then I'll open a PR.

Reply:
Yes it should; sorry, before the code was probably copy/pasted from a vector version where only part of the vector needed to be run, but the current code is wrong and your proposition correct.

Also, I just noticed that in the test . out there is actually the "TEST FAILED" but all those tests should PASS...should I push a PR with the patch or do we re-open this one?

Reply:
I have a patch already prepared. I will submit it soon.